Fine grained resource selection

knotseaborg opened the pull request:
I noticed that this issue hasn't been active for a month, so I'm giving it a try.

## Description

This draft addresses an enhancement suggestion for fine-grained selection of resources such as `GlobalNetworkSet`.

### Solution

I forced calico to set the label `projectcalico.org/kind` when a resource is created using the `clientv3` package. It seems to offer a consistent mechanism for resource selection.

### Tests

- Existing tests have been modified to verify the label after creation of a resource.
- Some tests in `calico/libcalico-go/lib/backend/syncersv1` have also been fixed by including the new label to prevent breakage.
---

@caseydavenport @sebhoss Let me know if this approach aligns with your intent, I can rework this draft.

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

## Related issues/PRs

<!-- If appropriate, include a link to the issue this fixes.
fixes <ISSUE LINK>

If appropriate, add links to any number of PRs documented by this PR
documents <PR LINK>
-->

fixes #9857

## Todos

- [x] Tests - Passed tests for `calico/libcalico-go`
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/calico?pullRequest=10763) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/calico?pullRequest=10763) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/calico?pullRequest=10763) it.</sub>

sebhoss commented:
I think it's great! I'm not quite sure how the selection of `clientv3` packages interacts with something like auto creation of host endpoints though. Does it use v3 as well or does it directly use v1 and therefore bypass the code in this PR?

CC @fasaxc

fasaxc commented:
Thanks for the PR.  IIUC, this will add the label on write to the datastore as long as you go through our client package.  TBH, I'm not sure if that's the best place to add the function because:

* It won't add the label to existing resources.  So, on upgrade, you'll have a mix of resources with and without it.
* It won't help folks who bypass our Go API and write CRDs directly.  Sadly there are quite a few folks who do this.
* It'll add the label to all our resources, not just the relevant ones.  This might cause datastore bloat or confuse controllers that try to reconcile labels.
* Someone malicious could change the label to bypass policy rules.

For the other labels that we auto-populate, we've tended to add them on read, when we convert datastore objects into our in-memory representation.  This has the downside that users can't see the labels with kubectl, but it does mean that we can easily retrofit a new special label and it prevents any sort of tampering.  I think I'd be inclined to do it that way (And then, if we later decide that making the labels "real" is important, we could do it for all the labels that we synthesize.)

It's not as generic, but perhaps the right place to change is the individual clients in libcalico-go/lib/backend/k8s/resources.

* WorkloadEndpoints are backed by pods, so there's a conversion function that's invoked that could add the label.
* The others are backed by a general-purpose CRD client, I might add an opt-in `applyKindLabel` flag to `customK8sResourceClient` that tells the generic client to add the extra label to those on read (need to cover the Get,List,Watch cases).

knotseaborg commented:
@fasaxc @sebhoss Thank you for the feedback. I am revising the draft PR and will push changes soon based on this suggestion.

> It's not as generic, but perhaps the right place to change is the individual clients in libcalico-go/lib/backend/k8s/resources.
> * WorkloadEndpoints are backed by pods, so there's a conversion function that's invoked that could add the label.
> * The others are backed by a general-purpose CRD client, I might add an opt-in applyKindLabel flag to customK8sResourceClient that tells the generic client to add the extra label to those on read (need to cover the Get,List,Watch cases).

fasaxc commented:
/sem-approve

knotseaborg commented:
Sorry, I didn't mean to close this draft (In my attempt to achieve a clean reset for the new approach, I accidentally closed it)

### Some updates

1. Commit 1 - `add kind label to customresource`
    * When `applyKindLabel` is enabled, `customk8sResourceClient` now adds the `projectcalico.org/kind` label to resources during Get, List, and Watch operations.

3. Commit 2 - `test kind label with globalnetworkset CRUD`
    * To validate this approach, I set `applyKindLabel=true` from the `NewGlobalNetworkSetClient` and updated related e2e tests.
    * Added label matching for `GlobalNetworkSet`-reads from the k8s datastore. I'm unsure if these test modifications fully align with expectations.

### Did the tests pass?

 No, but `libcalico-go` tests failed for both the `master` & this branch. The tests don't reach the part where updates can be validated. 

@fasaxc Could you please take a look at my commits to make sure that they align with your intent? Any feedback will be helpful.

knotseaborg commented:
@fasaxc Thank you for the feedback! I will work on implementing this approach.
> Perhaps you could call a common function from both dataplanes that checks if the resource is one of the kinds that we should label and, if so, does the labeling. You also need to tweak the workload endpoint client, which doesn't use the CRD client.

knotseaborg committed:
kind label add/del

knotseaborg committed:
modified k8s CRD client & workload endpoint

knotseaborg committed:
modified etcd client

knotseaborg committed:
included label in existing tests

caseydavenport commented:
Hey @knotseaborg @fasaxc, thanks for the work on this! I've been thinking about this feature some more and I think we should take a slightly different approach.

The label-on-read approach has a problem in v3 CRD mode, where users interact with the Kubernetes API directly and never go through our Go client. Policy selectors evaluate against what's stored in the datastore, so a synthesized label that only exists in-memory won't be visible to the selector engine in that mode.

Rather than trying to get the label stored reliably (which would need a mutating admission policy, migration logic for existing resources, etc), I think the cleaner solution is to add a new `kind` field directly on the rule's `source`/`destination` (the EntityRule type). Something like:

```yaml
apiVersion: projectcalico.org/v3
kind: GlobalNetworkPolicy
spec:
  ingress:
  - action: Allow
    source:
      kind: GlobalNetworkSet
      selector: "networking.internal/networkset == 'dns-servers'"
```

This avoids the label machinery entirely - Felix would filter the endpoint sets by kind when evaluating the rule. When `kind` is omitted, behavior is unchanged (matches all resource types, same as today). Allowed values would be `GlobalNetworkSet`, `NetworkSet`, `HostEndpoint`, `WorkloadEndpoint`, and/or `Pod`.

It's a bigger change than the label approach since it touches the API types and Felix's policy evaluation. What do you guys think? I'd be happy to help implement this if you think it's suitable.
